diff --git a/torch/distributed/__init__.py b/torch/distributed/__init__.py
index 5fb05a34771..d333360b83e 100644
--- a/torch/distributed/__init__.py
+++ b/torch/distributed/__init__.py
@@ -110,6 +110,8 @@ if is_available():
         _get_process_group_name,
     )
 
+    from .world_manager import WorldManager, C10dWorld
+
     from .rendezvous import (
         rendezvous,
         _create_store_from_options,
diff --git a/torch/distributed/distributed_c10d.py b/torch/distributed/distributed_c10d.py
index db553f6bf4a..3bb2d22e45f 100644
--- a/torch/distributed/distributed_c10d.py
+++ b/torch/distributed/distributed_c10d.py
@@ -1233,7 +1233,7 @@ def init_process_group(
     internals of c10d. This means we can ignore the value
     they provide as it not exposed in a public way.
     """
-    group_name = _process_group_name([], use_hashed_name=False)
+    # group_name = _process_group_name([], use_hashed_name=False)
     if backend == Backend.MPI:
         if world_size != -1 or rank != -1:
             warnings.warn(
